modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/SampleDataModelValidator.java

```diff
@@ -181,13 +206,95 @@ private String asGenericSignature(List<Type> types) {
+                                                      providerClass.getName(), component.getName(), asString(pageItemsType),
+                                                      asString(sampleDataPayloadType))));
+      return true;
+    }
+
+    List<Type> sampleDataCollectionGeneric = getInterfaceGenerics(sampleDataPayloadType, Collection.class);
+    if (sampleDataCollectionGeneric.isEmpty() || sampleDataCollectionGeneric.get(0) == null) {
+      reporter.addError(new Problem(component, format(
+                                                      "SampleDataProvider [%s] is used on component '%s' which is paged. The SampleDataProvider is thus expected to provide a payload of type 'Collection<%s>', but an unbounded Collection was found instead. Please provide the proper generic",
+                                                      providerClass.getName(), component.getName(), asString(pageItemsType))));
+      return true;
+    }
+
+    final Type sampleProviderCollectionType = sampleDataCollectionGeneric.get(0);
+
+    if (!pageItemsType.equals(sampleProviderCollectionType)) {
+      reporter.addError(new Problem(component, format(
+                                                      "SampleDataProvider [%s] is used on component '%s' which is paged. The SampleDataProvider is thus expected to provide a payload of type 'Collection<%s>', but a Collection<%s> was found instead.",
+                                                      providerClass.getName(), component.getName(), asString(pageItemsType),
+                                                      asString(sampleProviderCollectionType))));
+      return true;
+    }
+
+    String componentAttributesSignature = asString(outputGenericTypes.getSecond());
+    String providerAttributesSignature = asString(sampleDataProviderGenericTypes.getSecond());
+
+    if (!componentAttributesSignature.equals(providerAttributesSignature)) {
+      reporter.addError(new Problem(component, format(
+                                                      "SampleDataProvider [%s] is used on component '%s' which is paged. The SampleDataProvider is thus expected to provide attributes of type '%s' but it returns attributes of type '%s' instead",
+                                                      providerClass.getName(), component.getName(), componentAttributesSignature,
+                                                      providerAttributesSignature)));
+      return true;
+    }
+
+    return true;
+  }
+
+  private Pair<Type, Type> getPagingProviderGenerics(ConnectableComponentModel model) {
+    return model.getModelProperty(ImplementingMethodModelProperty.class)
+        .map(mp -> {
+          ParameterizedType type = (ParameterizedType) mp.getMethod().getGenericReturnType();
+          return new Pair<>(type.getActualTypeArguments()[0], type.getActualTypeArguments()[1]);
```

Comment:
Should we check the size of the array?

Reply:
you already know this will be a `PagingProvider`, thus you already know the array will have two positions.